### PR TITLE
eth, internal, les: get balance and call by block hash

### PR DESCRIPTION
**Pull request by markya0616**

### CODE REVIEW

File: eth/api_backend.go [hunk 1]
```diff
@@ -102,6 +102,16 @@ func (b *EthAPIBackend) StateAndHeaderByNumber(ctx context.Context, blockNr rpc.
 	return stateDb, header, err
 }
 
+func (b *EthAPIBackend) StateAndHeaderByHash(ctx context.Context, blockHash common.Hash) (*state.StateDB, *types.Header, error) {
+	// Otherwise resolve the block number and return its state
+	header, err := b.HeaderByHash(ctx, blockHash)
+	if header == nil || err != nil {
+		return nil, nil, err
+	}
+	stateDb, err := b.eth.BlockChain().StateAt(header.Root)
```
Comment (bailantaotao) [inline eth/api_backend.go:111]: Do we need to add sanity check to header?

File: internal/ethapi/api.go [hunk 2]
```diff
@@ -502,6 +503,17 @@ func (s *PublicBlockChainAPI) GetBalance(ctx context.Context, address common.Add
...
 
+// GetBalanceByHash returns the amount of wei for the given address in the state of the
+// given block hash. The rpc.LatestBlockNumber and rpc.PendingBlockNumber meta
+// block numbers are also allowed.
+func (s *PublicBlockChainAPI) GetBalanceByHash(ctx context.Context, address common.Address, blockHash common.Hash) (*hexutil.Big, error) {
+	state, _, err := s.b.StateAndHeaderByHash(ctx, blockHash)
+	if state == nil || err != nil {
+		return nil, err
+	}
+	return (*hexutil.Big)(state.GetBalance(address)), state.Error()
```
Comment (bailantaotao) [inline internal/ethapi/api.go:514]: ditto

File: eth/api_backend.go (hunk 1 above)
Comment (markya0616) [reply eth/api_backend.go:111]: Already checked if header is nil in https://github.com/getamis/go-ethereum/pull/180#discussion-diff-216921315R108.

File: internal/ethapi/api.go (hunk 2 above)
Comment (markya0616) [reply internal/ethapi/api.go:514]: Already checked if state is nil in https://github.com/getamis/go-ethereum/pull/180#discussion-diff-216921400R511.

**Review by bailantaotao — APPROVED** [review]

**Pull request merged**